File: src/corelibs/U2Designer/src/dashboard/StatisticsWidget.cpp
/**
 * UGENE - Integrated Bioinformatics Tools.
 * Copyright (C) 2008-2016 UniPro <ugene@unipro.ru>
 * http://ugene.net
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include <QTimer>

#include "StatisticsWidget.h"

namespace U2 {

inline static QString timeStr(qint64 timeMks) {
    QDateTime t;
    t.setTimeSpec(Qt::UTC);
    t.setMSecsSinceEpoch(timeMks / 1000);
    return t.toString("hh:mm:ss.zzz");
}

#if (QT_VERSION < 0x050400) //Qt 5.7
StatisticsWidget::StatisticsWidget(const QWebElement &container, Dashboard *parent)
#else
StatisticsWidget::StatisticsWidget(const QString &container, Dashboard *parent)
#endif
: TableWidget(container, parent)
{
    useEmptyRows = false;
    createTable();
    fillTable();
    connect(dashboard->monitor(), SIGNAL(si_workerInfoChanged(const QString &, const U2::Workflow::Monitor::WorkerInfo &)),
        SLOT(sl_workerInfoChanged(const QString &, const U2::Workflow::Monitor::WorkerInfo &)));
    connect(dashboard->monitor(), SIGNAL(si_updateProducers()),
        SLOT(sl_update()));
}

void StatisticsWidget::sl_workerInfoChanged(const QString &actor,
    const U2::Workflow::Monitor::WorkerInfo &info)
{
#if (QT_VERSION < 0x050400) //Qt 5.7
    updateRow(actor, createRowByWorker(actor, info));
#else
    //assert(0);
#endif
}

void StatisticsWidget::sl_update() {
    const WorkflowMonitor *m = dashboard->monitor();
    CHECK(NULL != m, );

    QMap<QString, WorkerInfo> infos = m->getWorkersInfo();
    foreach (const QString &actor, infos.keys()) {
#if (QT_VERSION < 0x050400) //Qt 5.7
        updateRow(actor, createRowByWorker(actor, infos[actor]));
#else
        dashboard->page()->runJavaScript("updateRow(\"" + container + "\", createRowByWorker(\"" + m->actorName(actor) + "\",\"" + timeStr(infos[actor].timeMks) + "\",\"" + QString::number(m->getDataProduced(actor)) + "\"), " + actor + ");");
#endif
    }
}

QList<int> StatisticsWidget::widths() {
    return QList<int>() << 40 << 30 << 30;
}

QStringList StatisticsWidget::header() {
    return QStringList() << tr("Element") << tr("Elapsed time") << tr("Output messages");
}

QList<QStringList> StatisticsWidget::data() {
    QList<QStringList> result;
    const WorkflowMonitor *m = dashboard->monitor();
    CHECK(NULL != m, result);
    QMap<QString, WorkerInfo> infos = m->getWorkersInfo();
    foreach (const QString &actorId, infos.keys()) {
#if (QT_VERSION < 0x050400) //Qt 5.7
        WorkerInfo &info = infos[actorId];
        QStringList row;
        row << actorId; // id
        row << createRowByWorker(actorId, info);
        result << row;
#else
        dashboard->page()->runJavaScript("updateRow(\"" + container + "\", createRowByWorker(\"" + m->actorName(actorId) + "\",\"" + timeStr(infos[actorId].timeMks) + "\",\"" + QString::number(m->getDataProduced(actorId)) + "\"), " + actorId + ");");
#endif
    }
    return result;
}

#if (QT_VERSION < 0x050400) //Qt 5.7
QStringList StatisticsWidget::createRowByWorker(const QString &actor, const WorkerInfo &info) {
    QStringList result;
    const WorkflowMonitor *m = dashboard->monitor();
    CHECK(NULL != m, result);

    result << wrapLongText(m->actorName(actor));
    result << timeStr(info.timeMks);
    result << QString::number(m->getDataProduced(actor));
    return result;
}
#endif

} // U2

File: src/corelibs/U2Designer/javascript/StatisticsWidget.js
/*WorkerInfo {
    QString actorId;
    QString actor;
    int ticks;
    qint64 timeMks;
    int countOfProducedData;
  };
*/
function StatisticsWidget(containerId){
  //create parent widget
  addWidget("<span lang=\"en\" class=\"translatable\">Common Statistics</span>" + "<span lang=\"ru\" class=\"translatable\">Общая статистика</span>", "overview_tab", 1, containerId);
  TableWidget.apply(this, arguments); //inheritance 
  //private
  var self = this;
  //protected
  this._useEmptyRows = false;
  //public
  this.widths = [40, 30, 30];
  this.headers = ["<span lang=\"en\" class=\"translatable\">Element</span>" + "<span lang=\"ru\" class=\"translatable\">Элемент</span>",
                  "<span lang=\"en\" class=\"translatable\">Elapsed time</span>" + "<span lang=\"ru\" class=\"translatable\">Прошедшее время</span>",
                  "<span lang=\"en\" class=\"translatable\">Output messages</span>" + "<span lang=\"ru\" class=\"translatable\">Выходные сообщения</span>"
                  ];
  
  this.sl_workerStatsInfoChanged = function(info){
    if (!self._isContainerExists()) {
      agent.sl_onJsError("Can't find container by id = " + self._containerId + "!");
      return;
    }
    self._updateRow(id(info.actorId), createRowByWorker(info));
  };
  this.sl_workerStatsUpdate = function(workersStatisticsInfo){
    if (!self._isContainerExists()) {
      agent.sl_onJsError("Can't find container by id = " + self._containerId + "!");
      return;
    }
    workersStatisticsInfo.forEach(function(workerInfo){
      self._updateRow(id(workerInfo.actorId), createRowByWorker(workerInfo));
    });
  };
  //private
  function createRowByWorker(info){
    var result = [];

    result.push(wrapLongText(/*m->actorName(*/info.actor));
    result.push(timeStr(info.timeMks));// << timeStr(info.timeMks);
    result.push(info.countOfProducedData);
    return result;
  }
  function timeStr(timeMks){
    var date = new Date(timeMks/1000);
    var milliseconds = date.getUTCMilliseconds() > 99 ? date.getUTCMilliseconds() : (date.getUTCMilliseconds() > 9 ? "0"+date.getUTCMilliseconds() : "00"+date.getUTCMilliseconds());
    var seconds = date.getUTCSeconds() > 9 ? date.getUTCSeconds() : "0"+date.getUTCSeconds();
    var minutes = date.getUTCMinutes() > 9 ? date.getUTCMinutes() : "0"+date.getUTCMinutes();
    var hours = date.getUTCHours() > 9 ? date.getUTCHours() : "0"+date.getUTCHours();
    return hours + ":" + minutes + ":" + seconds + "." + milliseconds;
  }
  function id(actor){
      return "stw_" + actor;
  }
  //constructor code
  this._createTable();
  this._fillTable();
  showOnlyLang(agent.lang); //translate labels
}
